File: src/RulesService.php
<?php

namespace TruFraudBot;

/**
 * Read/write rules from rules.txt (one rule per line). Used as context for fb (rules) chat.
 */
final class RulesService
{
    private string $path;

    public function __construct(?string $path = null)
    {
        $this->path = $path ?? (getcwd() . DIRECTORY_SEPARATOR . 'rules.txt');
    }

    public function getRulesFilePath(): string
    {
        return $this->path;
    }

    public function read(): string
    {
        if (!is_file($this->path)) {
            return '';
        }
        $content = @file_get_contents($this->path) ?: '';
        return trim($content);
    }

    /** @return list<string> */
    public function getLines(): array
    {
        $content = $this->read();
        if (trim($content) === '') {
            return [];
        }
        $lines = explode("\n", $content);
        $result = [];
        foreach ($lines as $line) {
            $line = trim($line);
            if ($line !== '') {
                $result[] = $line;
            }
        }
        return $result;
    }

    public function add(string $rule): void
    {
        $rule = trim($rule);
        if ($rule === '') {
            return;
        }
        $dir = dirname($this->path);
        if ($dir !== '' && !is_dir($dir)) {
            @mkdir($dir, 0755, true);
        }
        file_put_contents($this->path, $rule . "\n", FILE_APPEND | LOCK_EX);
    }

    public function remove(int $oneBasedIndex): bool
    {
        $lines = $this->getLines();
        if ($oneBasedIndex < 1 || $oneBasedIndex > count($lines)) {
            return false;
        }
        array_splice($lines, $oneBasedIndex - 1, 1);
        $content = $lines === [] ? '' : implode("\n", $lines) . "\n";
        file_put_contents($this->path, $content, LOCK_EX);
        return true;
    }

    public function edit(int $oneBasedIndex, string $newText): bool
    {
        $newText = trim($newText);
        $lines = $this->getLines();
        if ($oneBasedIndex < 1 || $oneBasedIndex > count($lines) || $newText === '') {
            return false;
        }
        $lines[$oneBasedIndex - 1] = $newText;
        file_put_contents($this->path, implode("\n", $lines) . "\n", LOCK_EX);
        return true;
    }
}

File: Services/RulesService.cs
namespace Truthbot.Services;

/// <summary>
/// Read/write rules from rules.txt (one rule per line). Used as context for fb (rules) chat.
/// </summary>
public class RulesService
{
    private readonly string _path;
    private static readonly object _lock = new();

    public RulesService(string? path = null)
    {
        // Use current working directory (project root when you "dotnet run") so rules.txt next to .csproj works
        _path = path ?? Path.Combine(Directory.GetCurrentDirectory(), "rules.txt");
    }

    public string RulesFilePath => _path;

    /// <summary>Full file content.</summary>
    public string Read()
    {
        if (!File.Exists(_path))
            return "";
        lock (_lock)
        {
            try
            {
                return File.ReadAllText(_path).Trim();
            }
            catch
            {
                return "";
            }
        }
    }

    /// <summary>Lines (non-empty).</summary>
    public IReadOnlyList<string> GetLines()
    {
        var content = Read();
        if (string.IsNullOrWhiteSpace(content))
            return [];
        return content.Split('\n', StringSplitOptions.RemoveEmptyEntries)
            .Select(l => l.Trim())
            .Where(l => l.Length > 0)
            .ToList();
    }

    /// <summary>Append one rule (one line).</summary>
    public void Add(string rule)
    {
        rule = rule.Trim();
        if (string.IsNullOrEmpty(rule))
            return;
        var dir = System.IO.Path.GetDirectoryName(_path);
        if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
            Directory.CreateDirectory(dir);
        lock (_lock)
        {
            File.AppendAllText(_path, rule + Environment.NewLine);
        }
    }

    /// <summary>Remove rule by 1-based index. Returns true if removed.</summary>
    public bool Remove(int oneBasedIndex)
    {
        var lines = GetLines().ToList();
        if (oneBasedIndex < 1 || oneBasedIndex > lines.Count)
            return false;
        lines.RemoveAt(oneBasedIndex - 1);
        lock (_lock)
        {
            File.WriteAllText(_path, lines.Count > 0 ? string.Join(Environment.NewLine, lines) + Environment.NewLine : "");
        }
        return true;
    }

    /// <summary>Replace rule at 1-based index with new text. Returns true if updated.</summary>
    public bool Edit(int oneBasedIndex, string newText)
    {
        newText = newText.Trim();
        var lines = GetLines().ToList();
        if (oneBasedIndex < 1 || oneBasedIndex > lines.Count || string.IsNullOrEmpty(newText))
            return false;
        lines[oneBasedIndex - 1] = newText;
        lock (_lock)
        {
            File.WriteAllText(_path, string.Join(Environment.NewLine, lines) + Environment.NewLine);
        }
        return true;
    }
}
